aiida/tools/groups/grouppaths.py

```diff
@@ -0,0 +1,321 @@
+                yielded.append(path_string)
+                try:
+                    yield GroupPath(
+                        path=path_string, type_string=self.type_string, warn_invalid_child=self._warn_invalid_child
+                    )
+                except InvalidPath:
+                    if self._warn_invalid_child:
+                        warnings.warn('invalid path encountered: {}'.format(path_string))  # pylint: disable=no-member
+
+    def __iter__(self):
+        # type: () -> Iterable[GroupPath]
+        """Iterate through all (direct) children of this path."""
+        return self.children
+
+    def __len__(self):
+        # type: () -> int
+        """Return the number of children for this path."""
+        return sum(1 for _ in self.children)
+
+    def __contains__(self, key):
+        # type: (str) -> bool
+        """Return whether a child exists for this key."""
+        for child in self.children:
+            if child.path_list[-1] == key:
+                return True
+        return False
+
+    def walk(self, return_virtual=True):
+        # type: () -> Iterable[GroupPath]
+        """Recursively iterate through all children of this path."""
+        for child in self:
+            if return_virtual or not child.is_virtual:
+                yield child
+            for sub_child in child.walk(return_virtual=return_virtual):
+                if return_virtual or not sub_child.is_virtual:
+                    yield sub_child
+
+    def walk_nodes(self, filters=None, node_class=None, query_batch=None):
+        # type: () -> Iterable[WalkNodeResult]
+        """Recursively iterate through all nodes of this path and its children.
+
+        Be aware that `query_batch` is only safe if no commit will take place during this transaction.
+        """
+        query = orm.QueryBuilder()
+        group_filters = {}
+        if self.path:
+            group_filters['label'] = {'or': [{'==': self.path}, {'like': self.path + self.delimiter + '%'}]}
+        if self.type_string is not None:
+            group_filters['type_string'] = self.type_string
+        query.append(orm.Group, filters=group_filters, project='label', tag='group')
+        query.append(
+            orm.Node if node_class is None else node_class,
+            with_group='group',
+            filters=filters,
+            project=['*'],
+        )
+        for (label, node) in query.iterall(query_batch) if query_batch else query.all():
+            yield WalkNodeResult(GroupPath(label, type_string=self.type_string), node)
+
+    @property
+    def browse(self):
+        """Return a ``GroupAttr`` instance, for attribute access to children."""
+        return GroupAttr(self)
+
+
+class GroupAttr:
+    """A class to provide attribute access to a ``GroupPath`` children.
+
+    The only public attributes on this class are dynamically created from the ``GroupPath`` child keys.
```

Comment:
So, yeah, these are dynamically created, however I would maybe clarify somewhere that these get frozen the moment you create the object. So, whereas the object `group_path` created by  `group_path = GroupPath()` will always reflect the current state of the database, the object `group_attr` created by `group_attr = group_path.browse` will only reflect the state of the database at the moment of its creation.

On the other hand, I am no expert in this, but is this the expected/desired behavior? Isn't it possible to run `c.path_list[-1]: c for c in self._group_path.children if REGEX_ATTR.match(c.path_list[-1])` inside of `__getattr__`? Or is this much more expensive / incompatible with autocomplete features?

Reply:
> will always reflect the current state of the database

Good point, yeh I think it's mainly on how this would affect autocompletion speed.
Probably need a bit of a manual test on that.

Reply:
Fixed in c8203d1